platform/pulp/tasking/system.py

```diff
@@ -0,0 +1,474 @@
+    else:
+        new_task = Task()
+        msg = _('The task status %(task_id)s exited immediately for some reason. Marking as '
+                'errored. Check the logs for more details')
+        runtime_exception = RuntimeError(msg % {'task_id': task_id})
+
+        class MyEinfo(object):
+            traceback = None
+
+        new_task.on_failure(runtime_exception, task_id, (), {}, MyEinfo)
+
+    ReservedResource.objects.get(task__id=task_id).delete()
+
+
+class UserFacingPulpTask(PulpTask):
+    """
+    A Pulp Celery task which will be visible to the user through the tasking portion of the API.
+
+    This object provides two interfaces to dispatch tasks: :meth: `apply_async` and
+    :meth: `apply_async_with_reservation`.
+
+    The :meth: `apply_asyc` provides normal celery dispatches of the task to the 'celery' queue
+    which all workers subscribe to. The task is handled by the first available worker.
+
+    The :meth: `apply_async_with_reservation` dispatch interface will send the task through the
+    resource_manager queue and process and will be assigned a specific worker. A series of
+    reservations cause specific types of tasks to not be run concurrently. See the
+    :meth: `apply_async_with_reservation` documentation for more details.
+    """
+
+    # this tells celery to not automatically log tracebacks for these exceptions
+    throws = (PulpCodedException,)
+
+    def apply_async_with_reservation(self, resource_type, resource_id, *args, **kwargs):
+        """
+        This method provides normal apply_async functionality, while also serializing tasks by
+        resource name. No two tasks that claim the same resource name can execute concurrently. It
+        accepts resource_type and resource_id and combines them to form a reservation key.
+
+        This does not dispatch the task directly, but instead promises to dispatch it later by
+        encapsulating the desired task through a call to a :func: `_queue_reserved_task` task. See
+        the docblock on :func: `_queue_reserved_task` for more information on this.
+
+        This method creates a :class: `pulp.platform.models.Task` object. Pulp expects to poll on a
+        task just after calling this method, so a Task entry needs to exist for it
+        before it returns.
+
+        For a list of parameters accepted by the *args and **kwargs parameters, please see the
+        docblock for the :meth: `apply_async` method.
+
+        :param resource_type: A string that identifies type of a resource
+        :type resource_type:  basestring
+
+        :param resource_id:   A string that identifies some named resource, guaranteeing that only
+                              one task reserving this same string can happen at a time.
+        :type  resource_id:   basestring
+
+        :param tags:          A list of tags (strings) to place onto the task, used for searching
+                              for tasks by tag
+        :type  tags:          list
```

Comment:
I added some docstrings indicating that `tags` and `group_id` are both pulled out of `kwargs`. You'll see the changes on my next push. At some point in the future I would like to improve the argument interface on this function but it is more than I want to do now.